### ISSUE TITLE
extend action-setfield for list handling

### ISSUE THREAD

**Issue by tobibeer** [Jermolene/TiddlyWiki5#1993]
As the functional overlap seems significant, implement a *list-mode* for the **action-setfield** widget as proposed here: https://github.com/Jermolene/TiddlyWiki5/issues/1989#issuecomment-145331905 with some or all of the following capabilities...

* add (default: to the end)...
    * `<$action-setfield $field="my-list" $add="foo"/>`
    * never allow duplicates, thus add only once / if not present
        * it's just the more common use case, that's why I'd use it as a default
* add to start...
    * `<$action-setfield $field="my-list" $add="foo" $at="0"/>`
* add after first...
    * `<$action-setfield $field="my-list" $add="foo" $at="1"/>`
* add before another...
    * `<$action-setfield $field="my-list" $add="foo" $before="bar"/>`
* add after another...
    * `<$action-setfield $field="my-list" $add="foo" $after="bar"/>`
* add even if already present, thus allowing duplicates
    * `<$action-setfield $field="my-list" $add-multiple="foo"/>`
* remove all from list...
    * `<$action-setfield $field="my-list" $remove="foo"/>`
* remove only first occurence...
    * `<$action-setfield $field="my-list"  $remove-first="foo"/>`
* remove only last occurence...
    * `<$action-setfield $field="my-list"  $remove-last="foo"/>`
* persist-sort a list-field...
    * `<$action-setfield $field="my-list"  $sort="<mode>"/>`
    * ascending: **sort**, **sortcs**, **nsort**, **nsortcs**
    * descending: **!sort**, **!nsort**, **!sortcs**, **!nsortcs**
* replace, meaning add **bar** only if **foo** was present and then deleted
    * `<$action-setfield $field="my-list" $replace="foo" $add="bar"/>`
* in general, also works with tags...
    * `<$action-setfield $field="tags" $add="foo"/>`

### Precedence ###

When any list action is defined, e.g. **$add**, **$remove**, **$replace**, etc... then **$value** is never set. The widget either handles the field as a whole or as a list field, not both at a time.

In this example, since **$add** triggers *list-mode*, **$value** would be silently ignored...
`<$action-setfield $field='baz' $add='foo' $value='bar'/>`

**Comment by inmysocks** [comment]
I have something that mostly works. The formatting done to the tag field that adds [[ and ]] isn't there but aside from that it seems good. It also doesn't have sort or replace implemented yet. It is also its own widget for now. I can combine it with action-setfield later.

So, how does the tags field handle the formatting where it adds [[ and ]] around tags with spaces? This implementation uses a comma separated list and I can't find where the tags field does anything differently.

Here is a demo:
http://ooktech.com/jed/ExampleWikis/ListManipulation/

**Comment by tobibeer** [comment]
@inmysocks, the problem with using **$item** is that you are missing another one when you want to do $action="replace".

I don't see the benefit of using **$item**, when **$add** is much more explicit.

**Comment by inmysocks** [comment]
The benefit of using $action and $item is that this way you don't have to give a title for remove, for example you can use <$action-setfield $action=remove $at=end/> It does of course have the problem with replace. Either way there is a problem, this is just the first way I thought of so if we come up with an idea to fix it using one or the other I can change it.

**Comment by Jermolene** [comment]
> So, how does the tags field handle the formatting where it adds [[ and ]] around tags with spaces? This implementation uses a comma separated list and I can't find where the tags field does anything differently.

@inmysocks the utility functions `$tw.utils.parseStringArray()` and `$tw.utils.stringifyList()` are used to convert between arrays and strings of double square bracket quoted titles.

**Comment by Jermolene** [comment]
@inmysocks in the case of the tags field, the field is registered in `$tw.Tiddler.fieldModules` by the boot kernel such that the field value is automatically converted to and from an array. We're deprecating that mechanism (@tobibeer may be able to provide a link to the recent discussion).

So, you'll need to deal with fields like "list" and "tags" that are already parsed into arrays, and other list fields that will have only been parsed as strings. You'll need to call parseStringArray() directly on those values to yield an array. Rather than checking explicitly for "list" and "tags", lookup the field in `$tw.Tiddler.fieldModules` to determine its type.

**Comment by felixhayashi** [comment]
> We're deprecating that mechanism (@tobibeer may be able to provide a link to the recent discussion).

@Jermolene but we will still be able to create a new tiddler programmatically and use 

```javascript
{
  title: "XYZ",
  tags: [ "foo", "bar" ]
}
```

instead of having to do this:

```javascript
{
  title: "XYZ",
  tags:  $tw.utils.stringifyList([ "foo", "bar" ])`
}
```

right? otherwise, it will break a lot of code and it would not be backwards compatible anymore. Same vice versa when *retrieving* a list.

Or did I misunderstand your comment?

-Felix

**Comment by inmysocks** [comment]
Thank you! That was driving me crazy. I updated the demo so it should be handled mostly correctly now. One problem is that as far as I can tell parseStringArray prevents you from having more than two identical entries on the list. I am sure there are plenty of other errors, but it should be the correct form at least.

**Comment by Jermolene** [comment]
> @Jermolene but we will still be able to create a new tiddler programmatically and use

Yes. By deprecating the mechanism, I mean:

* The core will not introduce any further tiddler field modules
* Instead, we'll treat fields polymorphically, coercing strings to lists and vice versa as required by particular operations

**Comment by danielo515** [comment]
It's funny how several of us are working on the same stuff.

I have extended the fieldmangler one week ago to fit my needs, but I don't need that features anymore.

Here is my code, it does not changes the fieldmangler in any way, it is just an extension:

https://gist.github.com/danielo515/bcdc4b79e9f82f29ecb4

As Jeremy said, I was using the parse strings utility functions to properly parse and set the fields.

@Jermolene , I'm afraid I don't understand at all what do you mean with deprecating.

> The core will not introduce any further tiddler field modules

I think that each plugin/module should be available to introduce their own field modules for the fields it is using. I though about it a couple of times

**Comment by felixhayashi** [comment]
> The core will not introduce any further tiddler field modules

I understand this part.

But

> Instead, we'll treat fields polymorphically, coercing strings to lists and vice versa as required by particular operations

But how would TW know how to parse or to stringify a field without having this somehow registered beforehand like in [boot.js](https://github.com/Jermolene/TiddlyWiki5/blob/c0c0b6653cd319721484fe272849b3788b0e4ea7/boot/boot.js#L824).

> I think that each plugin/module should be available to introduce their own field modules for the fields it is using. I though about it a couple of times

This is already possible if you create a tiddlerfield module. Or maybe I misunderstood your intention?

-Felix

**Comment by danielo515** [comment]
Hello @felixhayashi 
> This is already possible if you create a tiddlerfield module. Or maybe I misunderstood your intention?

Yes it is. I though about doing it since I have discovered recently, but I never did. I'm still learning about tiddly-wiki core! What I mean, is that the feature is very useful, and it should not be removed.

Regards

**Comment by inmysocks** [comment]
With what I linked to above any field can be used the same as a tags field. There is no reason to have the core treat the field differently because everything special can be done by telling the core how to treat the field when you want to manipulate it. 
The core adds [[ and ]] around tags with spaces in the current tags field, which is a hardcoded fundamental difference in how the fields are treated. The widget I made just reads what is in the field and treats it as a list, the content is processed and put into the field as a string. The core doesn't do anything special to the field. Doing it his way makes everything much more flexible and doesn't add more restrictions on what you can do, like now you can't have a field called 'tags' that does anything other than the default tag behaviour.

**Comment by felixhayashi** [comment]
Hi @danielo515 

> Yes it is. I though about doing it since I have discovered recently, but I never did. I'm still learning about tiddly-wiki core! What I mean, is that the feature is very useful, and it should not be removed.

I agree :) and I would not know how else to do it, I mean this is the standard way when programming in different layers (UI World \<-\> Object World \<-\> Database/Store World). You necessarily need to define beforehand how to parse and write values from the layer below…

@inmysocks 

We are talking about different things here. I was talking about *plugins* storing objects as strings and retrieving strings from the store as objects because I was concerned about Jeremy's remark about depreciating the tiddlerfields mechanism. So I wasn't talking about the user manipulating lists with an action-widget and I think it would be great to have such an action-widget that allows list manipulation. A similar issue is probably #1957.

-Felix

**Comment by inmysocks** [comment]
Ok, I misunderstood then. And I think that #1957 could be resolved using almost the same code I used here. If no one else does I will look at that later.

**Comment by Jermolene** [comment]
The primary purpose of the tiddlerfields mechanism was so to convert field values from the string values that are used for serialisation (ie saving or syncing) to the field values of the "Tiddler" object. Those field values are typed: one gets a string for the "text" and "title" fields, and an array of strings for the "tags" field.

The idea was that code that wanted to access the tags field as a list of tags didn't have to repeatedly convert the original string value into an array.

However, there have turned out to be several disadvantages with the mechanism:

* It doesn't work well with plugins that want to define the same field as different types
* It requires that a wiki be correctly configured to understand all the fields that will be used
* It doesn't help with the related problem of displaying fields with the most appropriate template or editor

So, the proposal going forwards is:

* We deprecate the tiddlerfields mechanism
 * That means that we keep it in the core, but we don't make any further changes to it, nor do we introduce any new tiddlerfield modules into the core
 * That means that we have to continue to cope with the fact that different tiddler fields will appear as different types
* New code that wishes to treat a field as a particular type should coerce the tiddler field values as required. In other words, if it is looking for a list and it obtains a string, it should use `$tw.utils.parseStringArray()` to convert the string into an array
* Perhaps we introduce a new convention based on tiddlers that allows the types of fields to be specified for the purposes of using the right editor or template

Is that a little clearer?

**Comment by tobibeer** [comment]
> Perhaps we introduce a new convention based on tiddlers that allows the types of fields to be specified for the purposes of using the right editor or template

I think it would be very good to have a basic standard convention / notation for how and where to define typed fields so that accessors in widgets / templates can be told to parse them and treat the content accordingly, rather than everyone specifying their very own way of where and how to store field-type-definitions.

For example, if I wanted a given field to be represented differently in edit-mode, it should be done by the field-rendering widgets to look up that definition and see that a field named `due` is to be represented by some date-picker plugin widget via some appropriate template routing.

**Comment by Jermolene** [comment]
> I think it would be very good to have a basic standard convention / notation for how and where to define typed fields so that accessors in widgets / templates can be told to parse them and treat the content accordingly, rather than everyone specifying their very own way of where and how to store field-type-definitions.

I think that proposal disrupts the layering. If we adopt a tiddler-based convention that would mean using that convention within templates to render fields in the correct way. Widgets should not need to be aware of the convention. If you want a convention that JavaScript widgets are aware of, then that's the existing tiddlerfields mechanism. In particular, the tiddlerfields mechanism **is** an attempt to allow widgets to avoid having to worry about parsing field content; we don't want to replace it with a mechanism designed on the same basis.

**Comment by tobibeer** [comment]
Yes. I am not necessarily talking about extending widgets to interpret any such configuration, but rather for templates to do the right routing. Whether that calls for extending widgets or would be more performant with changes on the widget layer is yet to be seen.

**Comment by felixhayashi** [comment]
Hi @Jermolene,

thanks for this clarification.

I stongly agree that no plugin or non-core code should be allowed to
register tiddlerfields as this might cause collisions! If that was your
point, then I fully agree! But, for core fields, I think autoboxing is a
great thing to have and I am grateful you implemented this as it saved
me from writing uncessary code for conversion when dealing with core
fields like "tags, list etc."

So I would not understand why you would want to completely depreciate it
(I mean also for fields "recognized/handled" by the core). Saying
something is depreciated means it is bad practice and shouldn't have
been used in the first place. But if I imagine that e.g. the "tags"
field were never handled by the core with autoboxing and everybody (who
writes code) needed to test for a string first and then do a conversion,
I would regard it as a huge overhead.

-Felix

**Comment by inmysocks** [comment]
Just as a note: I am cleaning up the widget a bit and I found a problem. $tw.utils.parseStringArray() only allows a single instance of each title in the list. I am not sure if changing the original function is a practical thing to do at this point so I made a version of the function that doesn't have this restriction in the widget. That is something we will need to address at some point.

**Comment by tobibeer** [comment]
@inmysocks, I think the most practical way would be to extend these core functions with a parameter that tells them not to remove duplicates, while keeping the current handling as default.

**Comment by Jermolene** [comment]
> But, for core fields, I think autoboxing is a great thing to have and I am grateful you implemented this as it saved me from writing uncessary code for conversion when dealing with core fields like "tags, list etc."

@felixhayashi there is another approach: the tiddler object could provide type-specific accessor methods that handle the conversion and caching. For example, `tiddler.fieldString["my-field-name"]`.

>  But if I imagine that e.g. the "tags"
field were never handled by the core with autoboxing and everybody (who
writes code) needed to test for a string first and then do a conversion,
I would regard it as a huge overhead.

We're not proposing to change the semantics of existing fields.

> Just as a note: I am cleaning up the widget a bit and I found a problem. $tw.utils.parseStringArray() only allows a single instance of each title in the list. I am not sure if changing the original function is a practical thing to do at this point so I made a version of the function that doesn't have this restriction in the widget. That is something we will need to address at some point.

I think we could consider changing it to prevent it removing duplicates. Perhaps open another ticket so we can discuss the implications? Alternatively, I'd agree with @tobibeer, we should consider adding an `options` parameter that contains a flag to control the duplicate handling.

**Comment by tobibeer** [comment]
:+1: 

> consider adding an options parameter that contains a flag to control the duplicate handling

**Comment by felixhayashi** [comment]
> @felixhayashi there is another approach: the tiddler object could provide type-specific accessor methods that handle the conversion and caching. For example, tiddler.fieldString["my-field-name"].

For directly accessing fields, this would be an option but it would also slow down access to field properties if you use JS accessor descriptors instead of parsing the fields in the beginning. Also if I initialise a `$tw.Tiddler` object with a field hashmap (`{ title:"xyz" … }`) I still would need to do conversion myself. I much prefer the way it is now: automatic and invisible.

> We're not proposing to change the semantics of existing fields.

I understood that. I am just arguing to not depreciate the mechanism for future core fields (if any are to come). But as said, I highly welcome the idea of depreciating tiddlerfield to be used by plugin authors as this might create conflicts. Maybe it could be allowed if they provided a namespace for the field (e.g.`pluginxyz.color`).

> I'd agree with @tobibeer, we should consider adding an options parameter that contains a flag to control the duplicate handling.

:+1:

**Comment by danielo515** [comment]
> there is another approach: the tiddler object could provide type-specific accessor methods that handle the conversion and caching. For example, tiddler.fieldString["my-field-name"].

That is what I was going to say. Why not extend the Tiddler object? currently the only use that I made of the Tiddler object is `fields=tiddler.fields`. I have wrote some wrappers around the tiddler object because it does not have many useful methods by default.

So we can end with something like
tiddler.tags.asString()
tiddler.tags.asArray()
tiddler.someDataField.asObject()

That would make it more semantically comprehensible.

Regards.

Regards.

**Comment by Jermolene** [comment]
Hi @danielo515 I think you're making a similar suggestion to mine above. The key point is that the conversion is cached within the tiddler object.

@felixhayashi your concerns about performance are what drove the original design, but my point is that now I've come to a different conclusion: that the original design was an example of premature optimisation, and that in fact the on the fly coercion is perfectly reasonable. It also matches how JavaScript works.

**Comment by danielo515** [comment]
Hello @Jermolene 
>Hi @danielo515 I think you're making a similar suggestion to mine above.

Indeed. It was my way of expressing my agreement, and that I have though about it also.

> that the original design was an example of premature optimisation, 

Ahh, that trap! I felt many times into it. But I never imagined a such talented programmer being victim of it. Now I fell less dumb.

Regards

**Comment by felixhayashi** [comment]
> Ahh, that trap! I felt many times into it. But I never imagined a such talented programmer being victim of it. Now I fell less dumb.

I know you are joking, but you should not feel dump in any case :) Also one can never prevent premature optimization because you never know what the future brings, so everybody can become a victim of that ;) Also I don't agree with Jeremy here, I think his tiddlerfields approach is very reasonable for core fields and using property accessors will not really help here when instantiating Tiddler objects as stringifiers would need to be called manually (see code below).

Just imagine the tiddlerfields mechanism wouldn't exist and you would have to do this everytime for core fields:

```javascript
$tw.wiki.addTiddler(new $tw.Tiddler({
  title: "XYZ",
  list: $tw.utils.stringifyList([ "1", "2", "3" ]),
  tags:  $tw.utils.stringifyList([ "foo", "bar" ]),
  modified: $tw.utils.stringifyDate(date),
}));
```

Anyhow there is no point in arguing, I mean undouptedly he knows best what is the best strategy for TW when it comes to core matters and since the mechanism will stay and not be removed but only be depreciated, I don't mind.

-Felix

**Comment by Jermolene** [comment]
> Just imagine the tiddlerfields mechanism wouldn't exist and you would have to do this everytime for core fields:

But that's not what is proposed; the tiddlerfields mechanism will be retained.

In any case, your example could still be done like this:

```
$tw.wiki.addTiddler(new $tw.Tiddler({
  title: "XYZ",
  list: [ "1", "2", "3" ],
  tags:  [ "foo", "bar" ],
  modified: date,
}));
```

In other words, the tiddler constructor would continue to coerce the provided fields to the type that is specified in the tiddlerfield module.

If an array were passed for a field that wasn't registered as a list then it would just be coerced into a string list.

**Comment by danielo515** [comment]
Dear @felixhayashi and @Jermolene 
I think that you disagree that you both agree :tongue: 

In any case @felixhayashi point would be valid if several new fields are added to the core:

```
$tw.wiki.addTiddler(new $tw.Tiddler({
  title: "XYZ",
  list: [ "1", "2", "3" ],
  tags:  [ "foo", "bar" ],
  modified: date,
  newcorefield: $tw.utils.stringifyNewCoreField("afafanfa:xxx:00_1")
}));
```

Regards

PD:
> I know you are joking, but you should not feel dump in any case :) 

Thank you mate ;)

**Comment by Jermolene** [comment]
> In any case @felixhayashi point would be valid if several new fields are added to the core:

No. If we added a new core field that didn't have a custom tiddlerfield module, then the field would be treated as a string, and any non-string value that you wanted to pass in would be automatically coerced into a string.

**Comment by felixhayashi** [comment]
> But that's not what is proposed

I am sorry, maybe Danielo is right an I am just misunderstanding what you are about to do. Where can I find the proposal?

> If an array were passed for a field that wasn't registered as a list then it would just be coerced into a string list.

:+1: would be great to do it like that. That would be even better than the current tiddlerfield mechanism which requires registration of field types before storing/retrieving. 

So for stringification you want to make it depend on the type that is injected using e.g.

* `instanceof Array` → $tw.utils.stringifyList()
* else `instanceof Date` →$tw.utils.stringifyDate()
* else `instanceof Object` → JSON.stringify()

In this case (or similar) I am totally d'accord! :)

**Comment by tobibeer** [comment]
@felixhayashi 

> I think his tiddlerfields approach is very reasonable for core fields

I also think that and that there can and should be a core method to `getCoreFields()` based on those modules. Because you might just be interested in those or then in "custom fields". Please see #1386, because I don't think we've quite arrived at a pragmatic solution yet, except for a minimalistic, manual listing of field-names to be indexed.

@Jermolene 

> If an array were passed for a field that wasn't registered as a list then it would just be coerced into a string list.

simple, reusable, :+1: 

@felixhayashi 

> So for stringification you want to make it depend on the type that is injected

dito :+1:

**Comment by matabele** [comment]
With the introduction of the $action-listops widget, it would appear that this issue can be closed. Any further requirements, especially with respect permitting multiple instances of items in lists can be discussed here: https://github.com/Jermolene/TiddlyWiki5/issues/2123

**Comment by Jermolene** [comment]
Thanks @matabele
